alf/bin/train.py

```diff
@@ -67,54 +71,149 @@ def _define_flags():
-    if trainer_conf.ml_type == 'rl':
-        trainer = policy_trainer.RLTrainer(trainer_conf)
-    elif trainer_conf.ml_type == 'sl':
-        trainer = policy_trainer.SLTrainer(trainer_conf)
-    else:
-        raise ValueError("Unsupported ml_type: %s" % trainer_conf.ml_type)
+    FLAGS.alsologtostderr = True
+    logging.set_verbosity(logging.INFO)
+
+    # When there are multiple processes, only the process with rank = 0 writes
+    # to the log directory to avoid race condition.
+    if rank == 0 and log_dir is not None:
+        logging.get_absl_handler().use_absl_log_file(log_dir=log_dir)
 
-    trainer.train()
+
+def _setup_device(rank: int = 0):
+    """Setup the GPU device for each process
+
+    Each process will use the GPU with the same rank (device id) by default.
+
+    Args:
+        rank: The ID of the process among all of the DDP processes
+    """
+    if torch.cuda.is_available():
+        alf.set_default_device('cuda')
+        torch.cuda.set_device(rank)
+
+
+def _snapshot_alf(root_dir: str):
+    """Snapshot the alf repository to the specified path
+
+    Args:
+        root_dir: the direcotry that stores various states of this training session
+    """
+    # ../<ALF_REPO>/alf/bin/train.py
+    file_path = os.path.abspath(__file__)
+    alf_root = str(pathlib.Path(file_path).parent.parent.parent.absolute())
+    # generate a snapshot of ALF repo as ``<root_dir>/alf``
+    common.generate_alf_root_snapshot(alf_root, root_dir)
+
+
+def training_worker(rank: int, world_size: int, conf_file: str, root_dir: str):
+    """An executable instance that trains and evaluate the algorithm
+    """
+    try:
+        _setup_logging(log_dir=root_dir, rank=rank)
+        _setup_device(rank)
+        print(world_size)
+        if world_size > 1:
+            # Specialization for distributed mode
+            dist.init_process_group('nccl', rank=rank, world_size=world_size)
+            # TODO(breakds): Remove this when DDP is finally working
+            # TODO(breakds): Also update the file level documentation when DDP is working
+            raise RuntimeError(
+                "Mutli-GPU DDP training is under development and temporarily unavailble"
+            )
+
+        common.parse_conf_file(conf_file)
+        trainer_conf = policy_trainer.TrainerConfig(root_dir=root_dir)
+
+        if trainer_conf.ml_type == 'rl':
+            trainer = policy_trainer.RLTrainer(trainer_conf)
+        elif trainer_conf.ml_type == 'sl':
+            # NOTE: SLTrainer does not support distributed training yet
+            if world_size > 1:
+                raise RuntimeError(
+                    "Multi-GPU DDP training does not support supervised learning"
+                )
+            trainer = policy_trainer.SLTrainer(trainer_conf)
+        else:
+            raise ValueError("Unsupported ml_type: %s" % trainer_conf.ml_type)
+
+        trainer.train()
+
+        # Load the configuration file, note that this will also bring up the
```

Comment:
unfinished sentence.

Reply:
Sorry about the mistake. I removed this sentence and put the complete comments where it should be.